### ISSUE TITLE
How to reload current collection with modified filters

### ISSUE THREAD

**Issue by NicoDG** [Smile-SA/elasticsuite#2448]
I'm looking for a way to reload the current collection with the active filters but with a different category filter.   
I was trying to clone the product collection and apply another category filter.

        $collection = clone $this->layer->getProductCollection();
        $collection->clear();
        $categoryId = '12';
        $categoryFactory = $this->categoryFactory->create();
        $category = $categoryFactory->load($categoryId);
        $collection->addCategoryFilter($category);
        $collection->load();
        $categorieFacets = $collection->getFacetedData('categories');

But the FacetedData doesn't seem to reload and returns the facets of the active filtering.
Could you point me in the right direction to force the facets to reload?

**Comment by NicoDG** [comment]
Basically i want to ignore the applied category when displaying the category filter.  That way, we will be able to show the full category tree which contains products that have a certain filter active.  

For example /men/jackets?color=yellow   I want to show the category tree of categories that contain products with the color yellow.  The only piece of the puzzle is being able to reload the faceteddata, because I see the correct aggregations being returned in the elastic results.

**Comment by rbayet** [comment]
Hello @NicoDG,

Do we agree that you expect to *change* the category filter, ie losing the current one ?
So in your exemple, you would show a category tree of categories containing yellow products, but not necessarily limited to the original list of products (ie visible somewhere under /men/jackets) ?

Regards,

**Comment by rbayet** [comment]
OK, the issue is that 
- the collection `$this->_isFiltersRendered` is not reset to false when the collection is clear()ed 
- calling `addCategoryFilter` does not reset it either

So when the collection is loaded again, renderFilterBefore is not called by \Magento\Framework\Data\Collection\AbstractDb::_renderFilters
```
    /**
     * Render sql select conditions
     *
     * @return  $this
     */
    protected function _renderFilters()
    {
        if ($this->_isFiltersRendered) {
            return $this;
        }

        $this->_renderFiltersBefore();
```

We added a past a fix for [a somehow similar issue](https://github.com/Smile-SA/elasticsuite/pull/1543) when the (new) sort order was not applied once the collection was loaded.

Logically, we should set `$this->_isFiltersRendered` to false every time a side effect to the potential Elasticsearch query is performed : 
- addFieldToFilter (since it's the inherited "DB collection" variant of \Magento\Framework\Data\Collection::addFilter which **does** reset it)
- addCategoryFilter
- addQueryFilter
- setSearchQuery
- setOrder (done)
- setPageSize
... but also all to all setters that have not been rewritten in the Elasticsuite collection but which imply that a new Elasticsearch query is to be constructed and executed through `_renderFiltersBefore()`. :|

@romainruaud what do you think ? Should we go that road or stick to the minimum approach of re-defining the `clear` method to add that ?
That could be enough for people cloning+clearing a collection before changing *some aspects* of it.
On the other hand, the [sort order issue](https://github.com/Smile-SA/elasticsuite/pull/1543) was due to a mishappening "pre-load" of the product list collection, so being exhaustive could make Elasticsuite a bit more bulletproof to theming errors or third party modules which prevent a filter or a sort order or a page size to be applied correctly, to the cost of extra Elasticsearch queries.

**Comment by rbayet** [comment]
Re-opening to check the generated query concerning the categories aggregation.

**Comment by rbayet** [comment]
Hello @NicoDG,

So, the problem was that the collection **was not** really reloaded and no new Elasticsearch query would be sent.
That has been fixed in 2.9.x and 2.10.x branches (specific `clear()` method), but you will need to add calls to new methods to do what you want.


With the fixed `clear()`, a new Elasticsearch query will be sent, but 
- if you don't have the Smile_ElasticsuiteVirtualCategory module enabled, the pre-existing category filter should be squashed as expected by calling `\Smile\ElasticsuiteCatalog\Model\ResourceModel\Product\Fulltext\Collection::addCategoryFilter`
- **but** if you have the Smile_ElasticsuiteVirtualCategory module enabled, the pre-existing category filter has been applied through `\Smile\ElasticsuiteVirtualCategory\Model\Layer\Filter\Category::applyCategoryFilterToCollection` and `\Smile\ElasticsuiteCatalog\Model\ResourceModel\Product\Fulltext\Collection::addQueryFilter` which creates a *query filter* that will not be overwritten by a subsequent call to `\Smile\ElasticsuiteCatalog\Model\ResourceModel\Product\Fulltext\Collection::addCategoryFilter` creates a simple *filter* 
   - this means that the generated query will actually have two categories clause : one for the category you're currently browsing/filtering on and one for the category you programatically added
   - so instead of "changing category", you add an extra condition : "products both in category1 AND category 2"
   - to fix that behavior, **you will need to call the new `removeQueryFilters()` on the collection**
- a lot of automated behaviours rely on a [context object](https://github.com/Smile-SA/elasticsuite/search?q=searchContext) which holds the current category / the current search query, you will need to do a alter that object otherwise 
   - the generated Elasticsearch query will still apply the `[blacklistedProducts](https://github.com/Smile-SA/elasticsuite/blob/e8e67b407f3f49a2bac16e3525173d19af9c0f7c/src/module-elasticsuite-catalog/etc/elasticsuite_search_request.xml#L25)` query filter based on the original category
   - the generated Elasticsearch will not contain the aggregations related to the layered navigation filters based on your new category (and probably only the default ones defined in XML: attribute_set_id, indexed_attributes)
   - that's not your objective, but if you want to get products sorted by category position (explicitely or by default), the "category.position" sort order will still look for the original category products positions

So, *with the merged fixes in 2.9.x and 2.10.x branches*, a working version of your code would look like this 
```
        /** @var Collection $collection */
        $collection = clone $this->getLayer()->getProductCollection();
        $collection->clear();
        // Required if Smile_ElasticsuiteVirtualCategory is enabled.
        $collection->removeQueryFilters();

        $categoryId = 20;
        $category = $this->categoryFactory->create()->load($categoryId);
        // So aggregations are correctly computed.
        $this->searchContext->setCurrentCategory($category);
        // So a standard category filter or a virtual rule filter is applied.
        $collection->addCategoryFilter($category);

        // Reproducing (if required) the category product positions.
        // See \Smile\ElasticsuiteCatalog\Plugin\LayerPlugin::setSortParams.
        $sortFilter = ['category.category_id' => $categoryId];
        $collection->addSortFilterParameters('position', 'category.position', 'category', $sortFilter);

        $categorieFacets = $collection->getFacetedData('categories');
```

PS: I used the categoryId 20 based on the Luma catalog, which is the "Women" category because it actually has children.

Regards,
